pandas/core/reshape.py

```diff
@@ -1153,6 +1175,48 @@ def get_dummies(data, prefix=None, prefix_sep='_', dummy_na=False,
     from pandas.tools.concat import concat
     from itertools import cycle
 
+    # Deprecate filling NaN values with zeros, GH15926
+    # When this is finally deprecated, simply remove this block
+    # of code and change the default to np.nan in the function signature
+    # of `get_dummies`.
+    if fill_value is None:
+        warnings.warn('The default behavior of filling NaN values '
+                      'with zeros will be deprecated. Use '
+                      '`df = pd.get_dummies(df, fill_value=0)` to reproduce '
+                      'this behavior', FutureWarning, 3)
+        fill_value = 0
+
+    # Infer the proper output dtype.
+    # GH15926
+    try:
+        if np.all(np.isfinite(data.values if hasattr(data, 'values') else data)):
+            any_null = True
+        else:
+            any_null = False
+    except TypeError:
+        any_null = False
```

Comment:
this block sucks. I am sure pandas has an internal for this. I will try and find it later.